modules/distributor/usage/tracker.go

```diff
@@ -0,0 +1,425 @@
+		v := make([]string, len(buffer))
+		copy(v, buffer)
+		b = &bucket{
+			// Metric description - constant for this pass now that the dimensions are known
+			descr:  prometheus.NewDesc("tempo_usage_tracker_bytes_received_total", "bytes total received with these attributes", t.sortedKeys, t.constLabels),
+			labels: v,
+		}
+		t.series[h] = b
+	}
+	return b
+}
+
+type Tracker struct {
+	mtx      sync.Mutex
+	name     string
+	tenants  map[string]*tenantUsage
+	labelsFn tenantLabelsFunc
+	maxFn    tenantMaxFunc
+	reg      *prometheus.Registry
+	cfg      PerTrackerConfig
+}
+
+func NewTracker(cfg PerTrackerConfig, name string, labelsFn tenantLabelsFunc, maxFn tenantMaxFunc) (*Tracker, error) {
+	u := &Tracker{
+		cfg:      cfg,
+		name:     name,
+		tenants:  make(map[string]*tenantUsage),
+		labelsFn: labelsFn,
+		maxFn:    maxFn,
+		reg:      prometheus.NewRegistry(),
+	}
+
+	err := u.reg.Register(u)
+	if err != nil {
+		return nil, err
+	}
+
+	go u.PurgeRoutine()
+
+	return u, nil
+}
+
+// getTenant must be called under lock.
+func (u *Tracker) getTenant(tenant string) *tenantUsage {
+	data := u.tenants[tenant]
+	if data == nil {
+		data = &tenantUsage{
+			series: make(map[uint64]*bucket),
+			constLabels: prometheus.Labels{
+				tenantLabel:  tenant,
+				trackerLabel: u.name,
+			},
+		}
+		u.tenants[tenant] = data
+	}
+	return data
+}
+
+func (u *Tracker) Observe(tenant string, batches []*v1.ResourceSpans) {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
+
+	dimensions := u.labelsFn(tenant)
+	if len(dimensions) == 0 {
+		// Not configured
+		// TODO - Should we put it all in the unattributed bucket instead?
+		return
+	}
+
+	max := u.maxFn(tenant)
+	if max == 0 {
+		max = u.cfg.MaxCardinality
+	}
+
+	var (
+		now                             = time.Now().Unix()
+		data                            = u.getTenant(tenant)
+		mapping, buffer1, buffer2, last = data.GetBuffersForDimensions(dimensions)
+	)
+
+	for _, batch := range batches {
+		unaccountedForBatchData, totalSpanCount := nonSpanDataLength(batch)
+
+		if totalSpanCount == 0 {
+			// Mainly to prevent a panic below, but is this even possible?
+			continue
+		}
+
+		// This is 1/Nth of the unaccounted for batch data that gets added to each span.
+		// Adding this incrementally as we go through the spans is the fastest method, but
+		// loses some precision. The other (original) implementation is to record span counts
+		// per series into a map and reconcile at the end. That method has more accurate data because
+		// it performs the floating point math once on the total, instead of accumulating 1/N + 1/N ... errors.
+		batchPortion := int(math.RoundToEven(float64(unaccountedForBatchData) / float64(totalSpanCount)))
+
+		// To account for the accumulated error we dump the remaining delta onto the first span, which can be negative.
+		// The result ensures the total recorded bytes matches the input.
+		firstSpanPortion := unaccountedForBatchData - batchPortion*totalSpanCount
```

Comment:
trying to figure out if there's any potential weird negatives here, but I got nothing. the first span should be basically random and so this should effectively randomly distribute the overflow. additionally i think most people will have resource only labels and so they will all end up in the same series anyway.